### ISSUE TITLE
use https on fishtest

### ISSUE THREAD

**Issue by vondele** [glinscott/fishtest#133]
as requested by @Stefano80 and discussed here

https://github.com/official-stockfish/Stockfish/issues/1267#issuecomment-360198514

Get rid of the warning (using firefox 57.0.4):
https://support.mozilla.org/en-US/kb/insecure-password-warning-firefox?as=u&utm_source=inproduct
I think using a let's encrypt certificate is the right idea.

**Comment by d3vv** [comment]
It is a hell for admins to upgrade ca-certificates on each server, -1 contributor minimum))

**Comment by ppigazzini** [comment]
From my research Waitress does not natively support SSL, so the solution should be [to setup a proxy server](https://docs.pylonsproject.org/projects/waitress/en/latest/#using-behind-a-reverse-proxy).

**Comment by tomtor** [comment]
@ppigazzini No solution for HTTPS but introducing concurrent request handling by simply running:

`gunicorn --paste production.ini --threads 4`

instead of pserve worked OK for me in small tests.

Just `sudo pip install gunicorn` did the job.

See: https://docs.pylonsproject.org/projects/pyramid_cookbook/en/latest/deployment/gunicorn.html

Could you try this on your http://94.198.98.239/

Would make adding a proxy server trivial and the single process setup makes efficient caching posible.

**Comment by ppigazzini** [comment]
@tomtor done, I also updated the test DB from official.
As reference:
1. it was also needed `sudo pip install futures`
2. I changed start.sh to continue to use systemd:
```bash
#!/bin/zsh
cd /home/fishtest/fishtest/fishtest
nohup gunicorn --paste production.ini --threads 4 > nohup.out&
```

**Comment by tomtor** [comment]
Except it should be:

`gunicorn --paste production.ini --threads 4 --worker-class gthread`

I corrected my comment (adding --worker-class ghtread) later, so my fault.

Without that clause you get one non threaded worker, according to the docs.

**Comment by tomtor** [comment]
@ppigazzini Do you plan to use this also on the real server in the near future?

**Comment by ppigazzini** [comment]
@tomtor development server updated.
This seems a low risk change: keep in mind that CentOS was never updated (anyway I fixed the yum broken configuration) and I ahave not the credential of the ISP control panel to administer the VM (backup, restart etc.)

The best option in long run is to have a fresh setup either with a brand new VM or with Docker.

**Comment by mcostalba** [comment]
Maybe it could be useful:

https://stackoverflow.com/questions/37717645/setup-gunicorn-with-lets-encrypt-cert

**Comment by tomtor** [comment]
@ppigazzini Any plans on implementing https soon? I think that using gunicorn as mentioned by @mcostalba is probably the easiest approach. An apache or nginx reverse proxy is also not very complex, but need probably a server OS software upgrade first. Any plans/schedule regarding the latter?

**Comment by ppigazzini** [comment]
@tomtor updated all the CentOS 6.8 packages (one by one), first step done (a little problem with pip ... mmm)

**Comment by ppigazzini** [comment]
Closed the duplicates https://github.com/glinscott/fishtest/issues/375 https://github.com/glinscott/fishtest/issues/335

**Comment by ppigazzini** [comment]
@tomtor I start testing your PR.

I successfully started nginx w/ 3 instances of pserve (ports 5000-5002) according to the pyramid example:
https://docs.pylonsproject.org/projects/pyramid_cookbook/en/latest/deployment/nginx.html

(but replacing `pserve production.ini\?http_port=5000` with `pserve production.ini http_port=5000`)

I rewrote the systemd unit file to start N servers according to:
https://serverfault.com/questions/730239/start-n-processes-with-one-systemd-service-file

**Comment by ppigazzini** [comment]
@tomtor the PR https://github.com/glinscott/fishtest/pull/383 is working fine on the dev server.

I have a problem w/ nginx load balancer:

W/ the default round robin mechanism the workers start the test but after few games stop w/ the message:
```
TC limit 484.8372013645258 End time: 2019-08-31 20:47:10.678155
Indexing opening suite...
Started game 1 of 250 (New-8fec883 vs Base-8fec883)
Started game 2 of 250 (Base-8fec883 vs New-8fec883)
Started game 3 of 250 (New-8fec883 vs Base-8fec883)
Finished game 1 (New-8fec883 vs Base-8fec883): 0-1 {Black wins by adjudication}
Score of New-8fec883 vs Base-8fec883: 0 - 1 - 0  [0.000] 1
Task updated successfully in 0.070326s
Server told us task is no longer needed
Wait 10 seconds before upload of PGN...
Uploading compressed PGN of 1308 bytes
Task exited

Fetch task...
Worker version checked successfully in 0.078121s
Task requested in 0.234381s
No tasks available at this time, waiting...

Fetch task...
Worker version checked successfully in 0.078118s
Task requested in 0.247961s
No tasks available at this time, waiting...

Fetch task...
Worker version checked successfully in 0.087309s
Task requested in 0.465977s
No tasks available at this time, waiting...
```

W/ the ip-hash mechanism that guarantee the persistence to the same server the workers run fine.

Any idea?

**Comment by tomtor** [comment]
The problem is that there is more than one python pyramid process running.
I should have noticed that earlier, sorry about that!

We may have only one because it maintains state internally, so the config
should not be:

upstream myapp-site {
    server 127.0.0.1:5000;
    server 127.0.0.1:5001;}


but:

upstream myapp-site {
    server 127.0.0.1:5000;}


And you should start a single Python pyramid in the systemd.config.

That should fix it.

The multiple threads and caches in Pyramid and fishtest code should handle
the load, and it does because we had recently over 10000 cores served in
the current setup :-)


Op za 31 aug. 2019 om 20:49 schreef ppigazzini <notifications@github.com>:

**Comment by ppigazzini** [comment]
@tomtor I supposed that the server had an internal status. The ip_hash mechanism works fine, but if the @noobpwnftw workers fleet share the same ip it's not that useful. I will look at the hash mechanism
http://nginx.org/en/docs/http/ngx_http_upstream_module.html#hash

```
upstream fishtest {
  ip_hash;
  server 127.0.0.1:6543;
  server 127.0.0.1:6544;
  server 127.0.0.1:6545;
}
```

**Comment by tomtor** [comment]
@ppigazzini The current fishtest server implementation can only work with a single server instance, any setup with more than one will fail in various ways.

Because each worker HTTP-request should update a single instance of the SPRT calculation for that specific task. Also handing out new tasks most be done from a single instance...

**Comment by tomtor** [comment]
@ppigazzini If you really have fun with configuring nginx you should find a configuration which:
 * routes each update_task/request_spsa/stop_run request with matching 'run_id' in the body to the same pserv instance. We could have a small pool of these pservs, eg 2 - 4.
* request_task/failed_task request to a single pserv instance which only handles these requests
* the other requests should be handed to a small pool with just handles info requests and PGN-uploads

Not sure if the above is sufficient or even correct, but it might result in a functional multi server fishtest.

Note that we do not *NEED* it, because github.com is currently the bottleneck, but it might be fun figuring this out ;-)

**Comment by ppigazzini** [comment]
@tomtor 
http://nginx.org/en/docs/http/load_balancing.html
* ip_hash
Specifies that a group should use a load balancing method where requests are distributed between servers based on client IP addresses. The first three octets of the client IPv4 address, or the entire IPv6 address, are used as a hashing key. The method ensures that requests from the same client will always be passed to the same server except when this server is unavailable. In the latter case client requests will be passed to another server. Most probably, it will always be the same server as well.
* hash
Specifies a load balancing method for a server group where the client-server mapping is based on the hashed key value. The key can contain text, variables, and their combinations.

**Comment by ppigazzini** [comment]
@tomtor the dev server is running fine with 3 pserve and ip_hash load balancing mechanism.
A change of the worker IP stops the test run, but this shouldn't be a catastrophic fail.

**Comment by tomtor** [comment]
@ppigazzini I am afraid we have a mis-communication, because all workers from *different IPs* should still access the same pserv when the 'run_id' in the request-body is the same. So ip_hash does NOT establish this. We would need to hash on the value of 'run_id' in the JSON-body, but I do not find anything that describes that in http://nginx.org/en/docs/http/load_balancing.html link.

Where is the description of "The key can contain text, variables, and their combinations", because that sounds interesting.

**Comment by ppigazzini** [comment]
@tomtor ops, I copyed the wrong link, here is
http://nginx.org/en/docs/http/ngx_http_upstream_module.html#hash

**Comment by ppigazzini** [comment]
@tomtor assumption: fishtest is working fine w/ only one pserve instance.

I found that every python process has poor concurrent performance because the GIL (Global Interpreter Lock)
https://groups.google.com/d/msg/pylons-discuss/kGNO3ifiacY/cX1scxxyAAAJ
https://hackernoon.com/has-the-python-gil-been-slain-9440d28fa93d

This explains why pylons suggested as best practice to load balancing a couple of pserve processes w/ only one backend server.

If I'm not wrong the Gunicorn workers are used only to handle the HTTP requests, so this could be useful to extract some parallelism fulfilling the run_id consistency.
https://medium.com/building-the-system/gunicorn-3-means-of-concurrency-efbb547674b7

**Comment by tomtor** [comment]
@ppigazzini 2 pserve instances is an option, one handles all viewing and one does all the updates.

You should route the .../view* requests to server1 and the rest to server 2, like

https://www.google.com/url?sa=t&source=web&rct=j&url=https://stackoverflow.com/questions/27220678/how-to-redirect-to-specific-upstream-servers-based-on-request-url-in-nginx&ved=2ahUKEwiWwfPKwbvkAhUFYVAKHdwnASAQFjARegQICxAB&usg=AOvVaw0UjrVadItNW9ldlCTKUhyt&cshid=1567750296467

**Comment by noobpwnftw** [comment]
Is it possible to use something like memcache or redis to store those global variables?
For a quick IPC mutex without language dependency I'd recommend a simple `flock` mechanism.

**Comment by tomtor** [comment]
creating a context switch (redis) for just updating a few python variables will take 100x longer than acquiring the GIL lock (which is similar to memcache because you still need a lock). Handling /tests in a separate pserv would gain something because building that page is CPU intensive.

**Comment by ppigazzini** [comment]
@tomtor w/ your knowledge of the server side, should it possible/easy to use the id_run parity to split the load of the workers?
 - a pserve process knows nothing about the id_run parity, must periodically checks if an id_run has changed the status to 'finished/stopped/purged etc'
 - the workers connect to a URI (test.stockfishchess.org/0 or test.stockfishchess.org/1) according to the id_run parity
 - the worker request for a new task should be split in these steps:
   - get a new id_run from a random pserve
   - use the run_id parity to reconnect to the right URI and get the id_run content

In this way we will interconnect the application and the network layers searching for a new level of points of failure :)

**Comment by tomtor** [comment]
> * the workers connect to a URI (test.stockfishchess.org/0 or test.stockfishchess.org/1) according to the id_run parity

Yes, this would be possible. But the gain would be small because in practice many workers will be working on the same test... When we have a huge number of workers the test queue will be empty and all will jump on the first test available.

> * get a new id_run from a random pserve

This cannot be a random pserve, because handing out new runs must be centralized. 

Some simple ways to maximize the number of workers which we can handle (although we are doing fine just now) would be to direct /tests, /users and /actions to a dedicated pserve, so these requests will not interfere with updating requests.

One step further would be to handle new test creation and handing out tasks and terminating them in a separate pserve, but this is a bit more tricky. I would be reluctant to change this when there is no real need. Mongodb is the real bottleneck when data must be persisted. Thats why I added all these caches to fishtest in 2018 because before that all pserve threads were just waiting on mongodb.

Using pypy for a significant speedup in general of the python code is also low hanging fruit.

**Comment by ppigazzini** [comment]
@tomtor using several pserve should create new mongodb connections improving preformance (as @noobpwnftw suggested in https://github.com/glinscott/fishtest/pull/180#issuecomment-374010546)

**Comment by ppigazzini** [comment]
@tomtor I'm testing this nginx configuration:
  - upstream w/ a single server for `/api/`
  - upstream load balancing multiple servers for `/api/upload_png` and everything else

Let me know if I'm wrong.

```
upstream backend_000 {
    server 127.0.0.1:6543;
    server 127.0.0.1:6544;
    server 127.0.0.1:6545;
    server 127.0.0.1:6546;
    server 127.0.0.1:6547;
}

upstream backend_001 {
    server 127.0.0.1:6548;
}

server  {
    server_name dfts-0.pigazzini.it;

    location / {
        proxy_pass http://backend_000;

        proxy_set_header X-Forwarded-Proto $scheme;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Host $host:$server_port;
        proxy_set_header X-Forwarded-Port $server_port;

        client_max_body_size    10m;
        client_body_buffer_size 128k;
        proxy_connect_timeout   60s;
        proxy_send_timeout      90s;
        proxy_read_timeout      90s;
        proxy_buffering         off;
        proxy_temp_file_write_size 64k;
        proxy_redirect          off;

        location /api/upload_pgn {
            proxy_pass http://backend_000;
        }

        location /api/ {
            proxy_pass http://backend_001;
        }

    }

    listen [::]:443 ssl ipv6only=on; # managed by Certbot
    listen 443 ssl; # managed by Certbot
    ssl_certificate /etc/letsencrypt/live/dfts-0.pigazzini.it/fullchain.pem; # managed by Certbot
    ssl_certificate_key /etc/letsencrypt/live/dfts-0.pigazzini.it/privkey.pem; # managed by Certbot
    include /etc/letsencrypt/options-ssl-nginx.conf; # managed by Certbot
    ssl_dhparam /etc/letsencrypt/ssl-dhparams.pem; # managed by Certbot

}


server  {
    if ($host = dfts-0.pigazzini.it) {
        return 301 https://$host$request_uri;
    } # managed by Certbot

  listen  80;
  listen [::]:80;
  server_name dfts-0.pigazzini.it;
    return 404; # managed by Certbot
}

```

**Comment by ppigazzini** [comment]
@tomtor mmm, the "Modify" button has not effect on priority and throughput also adding /tests/view/ at backend_001.

**Comment by tomtor** [comment]
@ppigazzini Looks good.

These /api could also be added to backend_000:

api_active_runs
get_run
download_pgn
download_pgn_100
request_version

5 instances for server_000 is probably a bit much, I suggest 2 to reduce the memory footprint.

**Comment by tomtor** [comment]
> @tomtor mmm, the "Modify" button has not effect on priority and throughput, still not working if adding /tests/view/ to backend_001.

You should also add /tests/test_* (wildcard URL) to backend_001 for modifying, deleting, etc

**Comment by ppigazzini** [comment]
@tomtor I'm trying to make working  /tests/{modify,delete,...}, I'll update my results here.

**Comment by ppigazzini** [comment]
@tomtor the server it's working fine w/ this configuration.
`location /tests` is mandatory and can be served from the load balancer backend, `/tests/view/` must be served from the reverse proxy backend (already done by `location /tests/`) otherwise, after a change of the priority/throughput values, the page shows different values w/ any browser refresh.
I'm using 5 servers for the load balancer to fail faster during configuration tests, the change it's very simple.

```
upstream backend_rp {
    server 127.0.0.1:6543;
}

upstream backend_lb {
    server 127.0.0.1:6544;
    server 127.0.0.1:6545;
    server 127.0.0.1:6546;
    server 127.0.0.1:6547;
    server 127.0.0.1:6548;
}

server {

    server_name  dfts-0.pigazzini.it;

    location / {
        proxy_pass http://backend_lb;

        proxy_set_header X-Forwarded-Proto $scheme;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Host $host:$server_port;
        proxy_set_header X-Forwarded-Port $server_port;

        client_max_body_size    10m;
        client_body_buffer_size 128k;
        proxy_connect_timeout   60s;
        proxy_send_timeout      90s;
        proxy_read_timeout      90s;
        proxy_buffering         off;
        proxy_temp_file_write_size 64k;
        proxy_redirect          off;

        location /api/active_runs {
            proxy_pass http://backend_lb;
        }

        location /api/download_pgn {
            proxy_pass http://backend_lb;
        }

        location /api/download_pgn_100 {
            proxy_pass http://backend_lb;
        }

        location /api/request_version {
            proxy_pass http://backend_lb;
        }

        location /api/upload_pgn {
            proxy_pass http://backend_lb;
        }

        location /api/ {
            proxy_pass http://backend_rp;
        }

        location /tests/user/ {
            proxy_pass http://backend_lb;
        }

        location /tests/ {
            proxy_pass http://backend_rp;
        }

        location /tests {
            proxy_pass http://backend_lb;
        }

    }

    listen [::]:443 ssl ipv6only=on; # managed by Certbot
    listen 443 ssl; # managed by Certbot
    ssl_certificate /etc/letsencrypt/live/dfts-0.pigazzini.it/fullchain.pem; # managed by Certbot
    ssl_certificate_key /etc/letsencrypt/live/dfts-0.pigazzini.it/privkey.pem; # managed by Certbot
    include /etc/letsencrypt/options-ssl-nginx.conf; # managed by Certbot
    ssl_dhparam /etc/letsencrypt/ssl-dhparams.pem; # managed by Certbot

}

server {
    if ($host = dfts-0.pigazzini.it) {
        return 301 https://$host$request_uri;
    } # managed by Certbot

    listen 80;
    listen [::]:80;

    server_name  dfts-0.pigazzini.it;
    return 404; # managed by Certbot
}
```

**Comment by tomtor** [comment]
@ppigazzini I suggest dedicating 1 specific pserve backend (backend_tests) ONLY for /tests because it caches the result for 2 seconds.

Distributing /tests over several pserve backends will introduce a large load on Mongodb, because the caching will be reduced by multiple /tests serving backends

**Comment by ppigazzini** [comment]
@tomtor done, also deleted the port 80 redirection to deal w/ both workers type (w/ and w/o https://github.com/glinscott/fishtest/pull/383). Changed the backend names :) 

```
upstream backend_runs {
    server 127.0.0.1:6543;
}

upstream backend_tests {
    server 127.0.0.1:6544;
}

upstream backend_lb {
    server 127.0.0.1:6545;
    server 127.0.0.1:6546;
}

server {
    listen 80;
    listen [::]:80;

    server_name  dfts-0.pigazzini.it;

    location / {
        proxy_pass http://backend_lb;

        proxy_set_header X-Forwarded-Proto $scheme;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Host $host:$server_port;
        proxy_set_header X-Forwarded-Port $server_port;

        client_max_body_size    10m;
        client_body_buffer_size 128k;
        proxy_connect_timeout   60s;
        proxy_send_timeout      90s;
        proxy_read_timeout      90s;
        proxy_buffering         off;
        proxy_temp_file_write_size 64k;
        proxy_redirect          off;

        location /api/active_runs {
            proxy_pass http://backend_lb;
        }

        location /api/download_pgn {
            proxy_pass http://backend_lb;
        }

        location /api/download_pgn_100 {
            proxy_pass http://backend_lb;
        }

        location /api/request_version {
            proxy_pass http://backend_lb;
        }

        location /api/upload_pgn {
            proxy_pass http://backend_lb;
        }

        location /api/ {
            proxy_pass http://backend_runs;
        }

        location /tests/user/ {
            proxy_pass http://backend_lb;
        }

        location /tests/ {
            proxy_pass http://backend_runs;
        }
        location /tests {
            proxy_pass http://backend_tests;
        }

    }

    listen [::]:443 ssl ipv6only=on; # managed by Certbot
    listen 443 ssl; # managed by Certbot
    ssl_certificate /etc/letsencrypt/live/dfts-0.pigazzini.it/fullchain.pem; # managed by Certbot
    ssl_certificate_key /etc/letsencrypt/live/dfts-0.pigazzini.it/privkey.pem; # managed by Certbot
    include /etc/letsencrypt/options-ssl-nginx.conf; # managed by Certbot
    ssl_dhparam /etc/letsencrypt/ssl-dhparams.pem; # managed by Certbot

}

**Comment by ppigazzini** [comment]
@tomtor here the situation some minutes after a reboot (top and push xcVm1t0EE). The load on mongodb is from `update_users.py`: I'm measuring the script time, it's seems very long...

```
top - 19:09:26 up 9 min,  2 users,  load average: 0.93, 0.86, 0.46
Tasks: 126 total,   1 running, 125 sleeping,   0 stopped,   0 zombie
%Cpu0  :   4.7/0.7     5[||||||                                                                                              ]
%Cpu1  :   3.6/1.0     5[|||||                                                                                               ]
%Cpu2  :  63.3/2.3    66[|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||                                   ]
%Cpu3  :  32.0/2.7    35[|||||||||||||||||||||||||||||||||||                                                                 ]
GiB Mem : 73.5/4.840    [|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||                           ]
GiB Swap:  0.0/10.000   [                                                                                                    ]

  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
    1 root      20       79704   8756   6624 S        0.2   0:01.18 /sbin/init splash
  237 root      19  -1   96900  13920  13224 S        0.3   0:00.12  `- /lib/systemd/systemd-journald
  289 root      20       47252   4452   3148 S        0.1   0:00.15  `- /lib/systemd/systemd-udevd
  412 systemd+  20      143972   3308   2776 S        0.1   0:00.05  `- /lib/systemd/systemd-timesyncd
  448 systemd+  20       72672   4992   4432 S        0.1   0:00.02  `- /lib/systemd/systemd-resolved
  510 daemon    20       30376   2232   2024 S        0.0            `- /usr/sbin/atd -f
  523 root      20      112524   3512   3188 S        0.1   0:00.02  `- /usr/sbin/irqbalance --foreground
  527 root      20      289588   6752   5868 S        0.1   0:00.02  `- /usr/lib/accountsservice/accounts-daemon
  532 root      20       33364   3216   2924 S        0.1            `- /usr/sbin/cron -f
  533 message+  20       52132   4064   3588 S        0.1   0:00.04  `- /usr/bin/dbus-daemon --system --address=systemd: --nofork --nopidfile --systemd-activation --syslog-only
  556 root      20       47272   5312   4772 S        0.1            `- /sbin/wpa_supplicant -u -s -O /run/wpa_supplicant
  558 root      20      202008  33520  13568 S        0.7   0:00.74  `- /usr/bin/python3 -Es /usr/sbin/firewalld --nofork --nopid
  561 root      20       72632   5908   5192 S        0.1   0:00.03  `- /lib/systemd/systemd-logind
  567 syslog    20      265080   4312   3544 S        0.1   0:00.01  `- /usr/sbin/rsyslogd -n
  572 root      20      172484  17300   9516 S        0.3   0:00.12  `- /usr/bin/python3 /usr/bin/networkd-dispatcher --run-startup-triggers
  576 root      20      290924   6420   5632 S        0.1            `- /usr/lib/policykit-1/polkitd --no-debug
  752 root      20      189328  20488  12552 S        0.4   0:00.10  `- /usr/bin/python3 /usr/share/unattended-upgrades/unattended-upgrade-shutdown --wait-for-signal
  756 mongodb   20     3097732 1.987g  37236 S 103.0 41.1   7:17.42  `- /usr/bin/mongod --config /etc/mongod.conf
  759 Debian-+  20       76304  12340   7792 S        0.2   0:00.20  `- /usr/sbin/snmpd -Lsd -Lf /dev/null -u Debian-snmp -g Debian-snmp -I -smux mteTrigger mteTriggerConf -f
  787 root      20       18224   1912   1784 S        0.0   0:00.03  `- /sbin/agetty -o -p -- \u --noclear tty1 linux
  801 root      20       74340   5708   4976 S        0.1            `- /usr/sbin/sshd -D
  887 root      20      110028   7128   6112 S        0.1   0:00.01      `- sshd: fishtest [priv]
 1079 fishtest  20      110028   3496   2480 S        0.1   0:00.05          `- sshd: fishtest@pts/0
 1080 fishtest  20       24712   4960   3292 S        0.1   0:00.05              `- -bash
 1188 fishtest  20       43956   3856   3120 R        0.1   0:01.14                  `- top
 1293 root      20      110028   7092   6076 S        0.1   0:00.01      `- sshd: fishtest [priv]
 1375 fishtest  20      110028   3684   2668 S        0.1                    `- sshd: fishtest@pts/1
 1376 fishtest  20       24712   5144   3476 S        0.1   0:00.04              `- -bash
 1426 fishtest  20      782904 290128  22236 S   4.0  5.7   0:43.54                  `- /home/fishtest/fishtest/fishtest/env/bin/python /home/fishtest/fishtest/fishtest/utils/update_users.py
  927 fishtest  20       78684   7576   6540 S        0.1   0:00.02  `- /lib/systemd/systemd --user
  931 fishtest  20      113688   2308     20 S        0.0                `- (sd-pam)
  963 root      20      144312   1684     60 S        0.0            `- nginx: master process /usr/sbin/nginx -g daemon on; master_process on;
  964 www-data  20      147204   9200   7144 S        0.2   0:01.17      `- nginx: worker process
  965 www-data  20      147392   9344   7196 S   1.0  0.2   0:01.84      `- nginx: worker process
  966 www-data  20      147200   9208   7144 S        0.2   0:01.79      `- nginx: worker process
  967 www-data  20      147404   9336   7196 S        0.2   0:01.66      `- nginx: worker process
  968 fishtest  20     1924180 135964  23044 S   3.0  2.7   0:15.88  `- /home/fishtest/fishtest/fishtest/env/bin/python /home/fishtest/fishtest/fishtest/env/bin/pserve production.ini http_port=6543
  969 fishtest  20     1854036 236412  21824 S        4.7   0:03.24  `- /home/fishtest/fishtest/fishtest/env/bin/python /home/fishtest/fishtest/fishtest/env/bin/pserve production.ini http_port=6545
  974 fishtest  20     1900100 413876  21948 S   0.3  8.2   0:04.12  `- /home/fishtest/fishtest/fishtest/env/bin/python /home/fishtest/fishtest/fishtest/env/bin/pserve production.ini http_port=6544
  975 fishtest  20     1926884 341780  21724 S        6.7   0:04.07  `- /home/fishtest/fishtest/fishtest/env/bin/python /home/fishtest/fishtest/fishtest/env/bin/pserve production.ini http_port=6546
    2 root      20                           S                      [kthreadd]
    3 root      20                           S              0:00.03  `- [ksoftirqd/0]
    5 root       0 -20                       S                       `- [kworker/0:0H]
    7 root      20                           S   0.3        0:00.69  `- [rcu_sched]
    8 root      20                           S                       `- [rcu_bh]
    9 root      rt                           S                       `- [migration/0]

```

**Comment by tomtor** [comment]
@ppigazzini Some indices are missing on Mongodb?

**Comment by ppigazzini** [comment]
@tomtor I opened a new issue https://github.com/glinscott/fishtest/issues/391 about `update_users.py` to keep clean this.
